avro4s-core/src/main/scala/com/sksamuel/avro4s/MagnoliaDerived.scala

```diff
@@ -0,0 +1,89 @@
+package com.sksamuel.avro4s
+
+import magnolia._
+
+import scala.language.experimental.macros
+import scala.reflect.runtime.universe._
+
+trait MagnoliaDerivedSchemaFors {
+  type Typeclass[T] = SchemaFor[T]
+
+  implicit def gen[T]: SchemaFor[T] = macro Magnolia.gen[T]
+
+  def dispatch[T: WeakTypeTag](ctx: SealedTrait[SchemaFor, T]): SchemaFor[T] = new ResolvableSchemaFor[T] {
+    def schemaFor(env: DefinitionEnvironment[SchemaFor], update: SchemaUpdate): SchemaFor[T] = {
+      env.get[T].getOrElse {
+        DatatypeShape.of(ctx) match {
+          case SealedTraitShape.TypeUnion => TypeUnions.schema(ctx, env, update)
+          case SealedTraitShape.ScalaEnum => SchemaFor[T](ScalaEnums.schema(ctx), DefaultFieldMapper)
+        }
+      }
+    }
+  }
+
+  def combine[T: WeakTypeTag](ctx: CaseClass[SchemaFor, T])(implicit fieldMapper: FieldMapper = DefaultFieldMapper): SchemaFor[T] = {
```

Comment:
I'm trying to understand why the field mapper was moved out of Encoder / Decoder into here. I guess its because it's only needed when generating certain encoders rather than all the time. Can you clarify for me?